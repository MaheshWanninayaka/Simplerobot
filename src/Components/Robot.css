.cell {
    width: 40px;
    height: 40px;
    display: inline-flex;
    justify-content: center;
    align-items: center;
    border: 1px solid #000;
    cursor: pointer;
    font-size: 20px;
}

.robot {
    background-color: #00f;
    color: #fff;
    border: 1px solid #00f;
}

.row {
    display: flex;
}

.grid {
    display: flex;
    flex-direction: column-reverse;
}

.controls {
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 20px;
}

.button-row {
    display: flex;
    justify-content: center;
    gap: 100px;
}

.arrow-button {
    background: #007bff;
    color: #fff;
    border: none;
    padding: 20px 40px;
    border-radius: 5px;
    font-size: 18px;
    cursor: pointer;
    margin-bottom: 20px;
}

    .arrow-button:hover {
        background: #0056b3;
    }

.robot-grid-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    text-align: center;
}

