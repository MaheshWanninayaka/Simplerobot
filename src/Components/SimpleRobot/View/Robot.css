.cell {
    width: 40px;
    height: 40px;
    display: inline-flex;
    justify-content: center;
    align-items: center;
    border: 1px solid #000;
    cursor: pointer;
    font-size: 20px;
}

.robot {
    background-color: #00f;
    color: #fff;
    border: 1px solid #00f;
}

.row {
    display: flex;
}

.grid {
    display: flex;
    flex-direction: column-reverse;
}

.controls {
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 20px;
}

.button-row {
    display: flex;
    justify-content: center;
    gap: 100px;
}

.arrow-button-up {
    background: #007bff;
    color: #fff;
    border: none;
    padding: 20px 40px;
    border-radius: 5px;
    font-size: 18px;
    cursor: pointer;
    margin-bottom: 20px;
}

    .arrow-button-up:hover {
        background: #0056b3;
    }

.arrow-button-down {
    background: #007bff;
    color: #fff;
    border: none;
    padding: 20px 40px;
    border-radius: 5px;
    font-size: 18px;
    cursor: pointer;
    margin-bottom: 20px;
}

    .arrow-button-down:hover {
        background: #0056b3;
    }

.arrow-button-right {
    background: #007bff;
    color: #fff;
    border: none;
    padding: 20px 40px;
    border-radius: 5px;
    font-size: 18px;
    cursor: pointer;
    margin-bottom: 20px;
}

    .arrow-button-right:hover {
        background: #0056b3;
    }

.arrow-button-left {
    background: #007bff;
    color: #fff;
    border: none;
    padding: 20px 40px;
    border-radius: 5px;
    font-size: 18px;
    cursor: pointer;
    margin-bottom: 20px;
}

    .arrow-button-left:hover {
        background: #0056b3;
    }


.robot-grid-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    text-align: center;
}
