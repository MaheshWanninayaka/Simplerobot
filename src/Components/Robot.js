import React, { useState } from "react";
import "./Robot.css";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faArrowUp, faArrowRight, faArrowDown, faArrowLeft } from "@fortawesome/free-solid-svg-icons";

const Robot = () => {
    const [position, setPosition] = useState({ x: 1, y: 1 });

    const move = (direction) => {
        const { x, y } = position;
        switch (direction) {
            case "North":
                setPosition({ x, y: Math.min(y + 1, 5) });
                break;
            case "East":
                setPosition({ x: Math.min(x + 1, 5), y });
                break;
            case "South":
                setPosition({ x, y: Math.max(y - 1, 1) });
                break;
            case "West":
                setPosition({ x: Math.max(x - 1, 1), y });
                break;
            default:
                break;
        }
    };

    const renderGrid = () => {
        const grid = [];
        for (let i = 5; i >= 1; i--) {
            const row = [];
            for (let j = 1; j <= 5; j++) {
                const isRobotHere = i === position.y && j === position.x;
                row.push(
                    <div
                        key={`${i}-${j}`}
                        className={`cell ${isRobotHere ? "robot" : ""}`}
                        onClick={() => setPosition({ x: j, y: i })}
                    >
                        {isRobotHere && "ðŸ¤–"} {/* Display a robot emoji */}
                    </div>
                );
            }
            grid.push(<div key={i} className="row">{row}</div>);
        }
        return grid;
    };

    return (
        <div>
            <h1>Robot Movement</h1>
            <div className="robot-grid-container">
                <p>Current Position: ({position.x}, {position.y})</p>
                <div>{renderGrid()}</div>
                <div className="controls">
                    <div className="button-row">
                        <button onClick={() => move("North")} className="arrow-button">
                            <div className="arrow-up">
                                <FontAwesomeIcon icon={faArrowUp} />
                            </div>
                        </button>
                    </div>
                    <div className="button-row">
                        <button onClick={() => move("West")} className="arrow-button">
                            <div className="arrow-left">
                                <FontAwesomeIcon icon={faArrowLeft} />
                            </div>
                        </button>
                        <button onClick={() => move("East")} className="arrow-button">
                            <div className="arrow-right">
                                <FontAwesomeIcon icon={faArrowRight} />
                            </div>
                        </button>
                    </div>
                    <div className="button-row">
                        <button onClick={() => move("South")} className="arrow-button">
                            <div className="arrow-down">
                                <FontAwesomeIcon icon={faArrowDown} />
                            </div>
                        </button>
                    </div>
                </div>
            </div>
        </div>
    );
};

export default Robot;