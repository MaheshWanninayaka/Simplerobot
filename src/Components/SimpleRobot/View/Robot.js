import React from "react";
import "./Robot.css";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faArrowUp, faArrowRight, faArrowDown, faArrowLeft } from "@fortawesome/free-solid-svg-icons";

export function Robot({ position, moveWithDelay, move }) {

    const renderGrid = () => {
        const grid = []; //added to put the rows of cells in the grid
        for (let i = 5; i >= 1; i--) {
            const row = [];
            for (let j = 1; j <= 5; j++) {
                const isRobotHere = i === position.y && j === position.x; // used to check the robot is in this cell location
                row.push(
                    <div
                        key={`${i}-${j}`}
                        className={`cell ${isRobotHere ? "robot" : ""}`}
                        data-testid={`cell-${j}-${i}`}
                        onClick={() => {
                            moveWithDelay(j, i);//added this to pass the correct direction for the click event
                        }}
                    >
                        {isRobotHere && "X"}
                    </div>
                );
            }
            grid.push(<div key={i} className="row">{row}</div>);//each row created with set of cells added to grid
        }
        return grid;
    };

    return (
        <div>
            <h1>Robot Movement</h1>
            <div className="robot-grid-container">
                <p>Current Position: ({position.x}, {position.y})</p>
                <div>{renderGrid()}</div>
                <div className="controls">
                    <div className="button-row">
                        <button onClick={() => move("North")} className="arrow-button-up" data-testid="arrow-button-up">
                            <div className="arrow-up" data-testid="arrow-up">
                                <FontAwesomeIcon icon={faArrowUp} />
                            </div>
                        </button>
                    </div>
                    <div className="button-row">
                        <button onClick={() => move("West")} className="arrow-button-left" data-testid="arrow-button-left">
                            <div className="arrow-left" data-testid="arrow-left">
                                <FontAwesomeIcon icon={faArrowLeft} />
                            </div>
                        </button>
                        <button onClick={() => move("East")} className="arrow-button-right" data-testid="arrow-button-right">
                            <div className="arrow-right" data-testid="arrow-right">
                                <FontAwesomeIcon icon={faArrowRight} />
                            </div>
                        </button>
                    </div>
                    <div className="button-row">
                        <button onClick={() => move("South")} className="arrow-button-down" data-testid="arrow-button-down">
                            <div className="arrow-down" data-testid="arrow-down">
                                <FontAwesomeIcon icon={faArrowDown} />
                            </div>
                        </button>
                    </div>
                </div>
            </div>
        </div>
    );
}
