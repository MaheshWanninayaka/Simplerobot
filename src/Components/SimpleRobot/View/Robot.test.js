import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { Robot } from './Robot';

describe('Robot Component', () => {
    const Move = jest.fn();
    const MoveWithDelay = jest.fn();
    const position = { x: 1, y: 1 };

    test('renders correctly', () => {
        const { container } = render(
            <Robot position={position} move={Move} moveWithDelay={MoveWithDelay} />
        );
        expect(container).toMatchSnapshot();
    });

    test('displays the current position', () => {
        render(
            <Robot position={position} move={Move} moveWithDelay={MoveWithDelay} />
        );

        expect(screen.getByText(`Current Position: (${position.x}, ${position.y})`)).toBeInTheDocument();
    });

    test('handles move functions correctly', () => {
        render(
            <Robot position={position} move={Move} moveWithDelay={MoveWithDelay} />
        );

        fireEvent.click(screen.getByTestId('arrow-button-up'));
        expect(Move).toHaveBeenCalledWith('North');

        fireEvent.click(screen.getByTestId('arrow-button-left'));
        expect(Move).toHaveBeenCalledWith('West');

        fireEvent.click(screen.getByTestId('arrow-button-right'));
        expect(Move).toHaveBeenCalledWith('East');

        fireEvent.click(screen.getByTestId('arrow-button-down'));
        expect(Move).toHaveBeenCalledWith('South');
    });

    test('handles moveWithDelay function correctly', () => {
        render(
            <Robot position={position} move={Move} moveWithDelay={MoveWithDelay} />
        );

        const cell = screen.getByTestId(`cell-${position.x}-${position.y}`);
        fireEvent.click(cell);
        expect(MoveWithDelay).toHaveBeenCalledWith(position.x, position.y);
    });
});
