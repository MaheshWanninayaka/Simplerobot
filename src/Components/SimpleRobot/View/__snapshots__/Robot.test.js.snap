// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Robot Component renders correctly 1`] = `
<div>
  <div>
    <h1>
      Robot Movement
    </h1>
    <div
      class="robot-grid-container"
    >
      <p>
        Current Position: (
        1
        , 
        1
        )
      </p>
      <div>
        <div
          class="row"
        >
          <div
            class="cell "
            data-testid="cell-1-5"
          />
          <div
            class="cell "
            data-testid="cell-2-5"
          />
          <div
            class="cell "
            data-testid="cell-3-5"
          />
          <div
            class="cell "
            data-testid="cell-4-5"
          />
          <div
            class="cell "
            data-testid="cell-5-5"
          />
        </div>
        <div
          class="row"
        >
          <div
            class="cell "
            data-testid="cell-1-4"
          />
          <div
            class="cell "
            data-testid="cell-2-4"
          />
          <div
            class="cell "
            data-testid="cell-3-4"
          />
          <div
            class="cell "
            data-testid="cell-4-4"
          />
          <div
            class="cell "
            data-testid="cell-5-4"
          />
        </div>
        <div
          class="row"
        >
          <div
            class="cell "
            data-testid="cell-1-3"
          />
          <div
            class="cell "
            data-testid="cell-2-3"
          />
          <div
            class="cell "
            data-testid="cell-3-3"
          />
          <div
            class="cell "
            data-testid="cell-4-3"
          />
          <div
            class="cell "
            data-testid="cell-5-3"
          />
        </div>
        <div
          class="row"
        >
          <div
            class="cell "
            data-testid="cell-1-2"
          />
          <div
            class="cell "
            data-testid="cell-2-2"
          />
          <div
            class="cell "
            data-testid="cell-3-2"
          />
          <div
            class="cell "
            data-testid="cell-4-2"
          />
          <div
            class="cell "
            data-testid="cell-5-2"
          />
        </div>
        <div
          class="row"
        >
          <div
            class="cell robot"
            data-testid="cell-1-1"
          >
            X
          </div>
          <div
            class="cell "
            data-testid="cell-2-1"
          />
          <div
            class="cell "
            data-testid="cell-3-1"
          />
          <div
            class="cell "
            data-testid="cell-4-1"
          />
          <div
            class="cell "
            data-testid="cell-5-1"
          />
        </div>
      </div>
      <div
        class="controls"
      >
        <div
          class="button-row"
        >
          <button
            class="arrow-button-up"
            data-testid="arrow-button-up"
          >
            <div
              class="arrow-up"
              data-testid="arrow-up"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-arrow-up "
                data-icon="arrow-up"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 384 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M214.6 41.4c-12.5-12.5-32.8-12.5-45.3 0l-160 160c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L160 141.2V448c0 17.7 14.3 32 32 32s32-14.3 32-32V141.2L329.4 246.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3l-160-160z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </button>
        </div>
        <div
          class="button-row"
        >
          <button
            class="arrow-button-left"
            data-testid="arrow-button-left"
          >
            <div
              class="arrow-left"
              data-testid="arrow-left"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-arrow-left "
                data-icon="arrow-left"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.2 288 416 288c17.7 0 32-14.3 32-32s-14.3-32-32-32l-306.7 0L214.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </button>
          <button
            class="arrow-button-right"
            data-testid="arrow-button-right"
          >
            <div
              class="arrow-right"
              data-testid="arrow-right"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-arrow-right "
                data-icon="arrow-right"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M438.6 278.6c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L338.8 224 32 224c-17.7 0-32 14.3-32 32s14.3 32 32 32l306.7 0L233.4 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l160-160z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </button>
        </div>
        <div
          class="button-row"
        >
          <button
            class="arrow-button-down"
            data-testid="arrow-button-down"
          >
            <div
              class="arrow-down"
              data-testid="arrow-down"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-arrow-down "
                data-icon="arrow-down"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 384 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M169.4 470.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 370.8 224 64c0-17.7-14.3-32-32-32s-32 14.3-32 32l0 306.7L54.6 265.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
